### PR TITLE
fix(api): bound retries by wall-clock budget, not just attempt count

### LINKED ISSUE #97: Requests to Legacy REST endpoints failing
I am trying to use unifi-mcp-server with Claude and it is failing. I requested Claude to document the error and its conditions. It provided detailed data and suggested fixes
## Bug Description

Requests to legacy REST endpoints (list_port_forwards, get_device_connections) hang for ~4 minutes instead of failing fast, due to stacked retry/backoff on authenticate() + data call

## Error Message

```
[Paste the exact error message here]
```

## Environment Details

unifi-mcp-server version: 0.2.6 (ghcr.io/enuno/unifi-mcp-server:latest, image built 2026-05-04)
Deployment: Docker container, UNIFI_API_TYPE=local
Controller: UniFi Dream Machine Pro (UDM Pro), current firmware
Auth: local Network Application API key (Control Plane > Integrations)
MCP client: Claude (Claude Desktop / claude.ai), 4-minute client-side tool-call timeout

## Expected Behavior

The tool should execute successfully and return valid data without errors.

## Actual Behavior

Several tools that hit local-gateway legacy REST endpoints (/proxy/network/api/s/{site}/rest/...) hang indefinitely from the client's perspective — long enough to exceed a typical MCP client's ~4-minute tool-call timeout — rather than failing fast or returning a clear timeout error. Other endpoints on the modern Integration API (list_sites, get_site_details, list_vlans) generally respond quickly, though they occasionally show the same one-off slow-then-fast pattern on a cold connection.

## Additional Context

Reproduction
1. Configure the server against a UDM Pro with UNIFI_API_TYPE=local and a valid local Network API key.
2. Call list_port_forwards(site_id="default") or get_device_connections(device_id=None, site_id="default").
3. Observe: no response, no error, for the full client-side timeout window (in our case, consistently ~4 minutes, reproduced 3/3 attempts for each tool). The container itself remains healthy throughout (health checks passing, no crash, no restart).

## Possible Root Cause

Root cause (from reading src/api/client.py)
1. _request() retries on httpx.TimeoutException with exponential backoff: max_retries=3 (default), retry_backoff_factor=2.0 (default), request_timeout=30s (default). Worst case per call: ~4 attempts x 30s + backoff sleeps.
2. Tools such as list_port_forwards open a new UniFiClient and call await client.authenticate() before the actual data request (see src/tools/port_forwarding.py). authenticate() itself calls _request() (GET /ea/sites), so it has its own independent retry/backoff cascade.
3. When the controller is slow to answer a particular endpoint, the two cascades stack: authenticate() retries can consume up to ~2 minutes, then the actual data call retries can consume another ~2 minutes, totaling close to 4 minutes — which lines up almost exactly with common MCP client-side timeouts, making the server look completely unresponsive when it is technically still working through its own retry logic and would eventually raise a clean NetworkError.

- [ ] Authentication/Authorization issue
- [X ] API endpoint not available in this environment
- [ ] SSL certificate validation problem
- [X ] Rate limiting or timeout
- [ ] Invalid parameter validation
- [ ] Response format parsing error
- [ ] Other: [describe]

## Impact Assessment

Any MCP client with a timeout shorter than the worst-case stacked retry duration (roughly 2x max_retries x request_timeout, plus backoff) will report these tools as fully hung/unresponsive, even though the server is functioning as designed — just too slowly for practical interactive use.

**Severity**: [ ] Critical (blocks all operations) | [X ] High (major functionality broken) | [ ] Medium (workaround available) | [ ] Low (minor inconvenience)

**Affected Users**: [X] All users | [ ] Cloud API users only | [ ] Local gateway users only | [ ] Specific configuration only

**Workaround Available**: [ ] Yes | [X] No

Suggested fixes

1. Cache/reuse the authenticated client (or at least the auth state) across tool calls within a session instead of re-authenticating via a full retriable request on every single tool invocation.
2. Make request_timeout / max_retries configurable per-call or lower the defaults for read-only list/get operations, so a slow endpoint fails fast with a clear error rather than silently retrying past common client-side timeout windows.
3. Surface partial-failure/timeout state back to the MCP client sooner (e.g., an intermediate progress notification) rather than only returning after all retries are exhausted.
4. Consider whether list_port_forwards / get_device_connections specifically need investigation against real UDM Pro hardware — they failed 3/3 while list_vlans (also a /rest/ legacy endpoint) succeeded on retry, suggesting these two may have an additional, endpoint-specific slowness on the controller side worth documenting even if the retry-stacking fix above is implemented.

## Testing Notes

---

**Checklist for Manual Bug Reports**:

- [ ] Error message included
- [ X] Environment details provided
- [ X] Steps to reproduce documented
- [ X] Expected vs actual behavior described
- [ X] Logs or test results attached (if available)
- [ X] Labels applied: `bug`, `needs-triage`, `mcp-testing`

### PR DESCRIPTION
**Pull request by jpennin5**
## Description

**Problem:** Issue #97 reports tools hanging for minutes before failing.

**Root cause:** with the default 30 s request timeout and 3 retries, one
logical request can legitimately burn ~127 seconds (4 attempts plus
exponential backoff), and a tool that authenticates before its data call
stacks two of those — the reported four-minute hangs. That is longer than
most MCP clients' tool-call timeout, so the failure presents as tools that
never return rather than tools that fail.

**Fix:** retries now also respect a total wall-clock budget
(`UNIFI_RETRY_TOTAL_TIMEOUT`, default 60 s) covering all attempts and
backoff waits for one request. The check runs before each backoff sleep
(429, timeout, and network-error paths), so exhaustion fails fast with a
message naming the budget and the environment knob rather than sleeping
into a timeout the caller experiences as a hang. The attempt-count limit
is unchanged; the budget is an additional bound.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)

Default behavior only changes when retries would previously have slept
past 60 s of wall clock; the budget is configurable.

## Testing

- Unit tests cover: budget exhaustion raising `NetworkError` with the
  knob named in the message, the budget check running before (not after)
  a sleep, and the attempt counter still applying within a generous
  budget.
- The hang itself is a timing pathology and is exercised via the budget
  arithmetic rather than real sleeps.
- `pytest tests/unit/` and `pre-commit run --all-files` pass.
- `.env.example` / `.env.docker.example` document the new variable.

Closes #97

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (CHANGELOG.md, env examples)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published (none — stands alone)

## AI Assistance

This PR was created with assistance from Claude Code.

**Human Review Status:** ✅ Reviewed and approved by @jpennin5
**Testing:** Full unit suite passing, including new retry-budget tests.
**Security Review:** Completed — no credentials or identifying values;
configuration via environment variable.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -47,6 +47,7 @@ MCP_LOG_LEVEL=INFO
 # Optional: Advanced Settings
 # UNIFI_TIMEOUT=30
 # UNIFI_MAX_RETRIES=3
+# UNIFI_RETRY_TOTAL_TIMEOUT=60
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply .env.example:50]: Fixed in a224330 — both env examples now name `UNIFI_REQUEST_TIMEOUT`, matching the settings alias. Pre-existing, but the file was already in this PR's blast radius.

**Comment by enuno** [comment]
### Summary
Request changes. The PR is not yet merge-ready because the current status checks are not clean.

### Critical findings
- Merge state: \ / \.
- Failing check(s): Mermaid Diagram Sync Assistant.
- The changed tests surface should be revalidated after the failing check is resolved.

### Suggestions
1. Resolve the failing status check(s) and rerun CI.
2. Reconfirm the changed tests behavior against the intended controller/fixture shapes.
3. Ensure regression tests cover the original failure mode and edge cases introduced here.
4. Update docs if any request/response or error semantics changed.

### Verification checklist for the author
- [ ] Required checks are passing
- [ ] Regression tests cover the changed path
- [ ] Lint / format / type checks pass locally
- [ ] Documentation is updated where user-facing behavior changed
- [ ] No secrets, credentials, or controller-identifying data were introduced

**Comment by jpennin5** [comment]
The failing check is the `Mermaid Diagram Sync Assistant` GitHub App, not a workflow in this repository. Its check-run output on this PR reads: "Webhook Processing Failed — An unexpected error occurred while processing the webhook. Please try pushing a new commit or check your repository configuration." That processing happens on the app's side; pushing new commits to these PRs has re-triggered the check with the same result, so it is not reachable from a fork.

This is the loop documented in #112: the app has errored on every fork PR since late June, fork CI runs sit at `action_required` awaiting maintainer approval, and the daily automated review then reads the app error as a failing required check and requests changes.

Every check that can conclude on a fork PR is green here (both Socket Security checks). The remaining steps are repo-side — approving the CI runs and either repairing/removing the app integration or unrequiring its check. #117 merged with the identical check state, so the merge path exists. Happy to rebase or adjust anything code-side the moment CI can actually run.
